src/Microsoft.AspNetCore.Server.Kestrel.Https/Internal/HttpsConnectionAdapter.cs

```diff
@@ -110,8 +113,43 @@ public Task<IAdaptedConnection> OnConnectionAsync(ConnectionAdapterContext conte
 
             try
             {
-                await sslStream.AuthenticateAsServerAsync(_serverCertificate, certificateRequired,
-                        _options.SslProtocols, _options.CheckCertificateRevocation);
+                if (AppContext.TryGetSwitch(DisableHandshakeTimeoutSwitch, out var handshakeDisabled) && handshakeDisabled)
+                {
+                    await sslStream.AuthenticateAsServerAsync(
+                        _serverCertificate, certificateRequired, _options.SslProtocols, _options.CheckCertificateRevocation);
+                }
+                else
+                {
+                    try
+                    {
+                        var handshakeTask = sslStream.AuthenticateAsServerAsync(
```

Comment:
Should we pass in a cancellation token anyways? You can imagine a scenario where this is fixed in .NET Core 2.1 where people run ASP.NET Core 1.x on .NET Core 2.1. It would just work like magic 😄

Reply:
Yes, they are adding cancellation support.